### PR TITLE
remove synthetic status results

### PR DESCRIPTION
**Pull request by cforkish**
this PR cleans up some tech debt affecting the behavior and display of our progress bubbles, as a precursor to cleaning up our bubble components.

## readonly
removed:
* `TestResults.READONLY_SUBMISSION_RESULT`
* `LevelStatus.readonly`

it turns out that our progress bubbles don't ever need this value from `UserLevel`. previously, we were showing bubbles as completed if they were `readonly`, but now that we've untangled lock state from submission status and are able to display actual progress status when levels are locked, we want to do the same for `readonly`. that means that bubbles in the `readonly` state are identical to those in the `unlocked` state, and consequently we no longer need this value on the frontend.

## locked
removed:
* `TestResults.LOCKED_RESULT`
* `LevelStatus.locked`

now that we are using `levelType.isLocked` and `studentLevelProgressType.locked` to determine a level's lock state for a user, having that state also encoded in `status` and `result` values is both unnecessary and confusing. both have been removed, and anywhere they were being used has been updated to look at the actual lock state instead.

on a related note, previously `levelStatus === LevelStatus.locked` was being used to determine if `ProgressTableLevelBubble` should be disabled. however, it turns out that bubbles in the progress table should _never_ be disabled since teachers should be able to click them even if they're locked. consequently, `ProgressTableLevelBubble` has been updated to no longer support the disabled case, since starting with the upcoming #40193, `ProgressTableLevelBubble` is being formalized as a table-specific bubble variant.

additionally, not needing to handle teachers as a special case for lock state means we can remove the `user.authorized_teacher?` check from `UserLevel.show_as_locked?` and significantly simplify the logic in `users_helper.merge_user_progress_by_level`.

## Links
- jira ticket: [LP-1865]

[LP-1865]: https://codedotorg.atlassian.net/browse/LP-1865

### CODE REVIEW

File: apps/src/templates/sectionProgress/progressTables/ProgressTableDetailCell.jsx [hunk 1]
```diff
@@ -84,7 +82,6 @@ export default class ProgressTableDetailCell extends React.Component {
             >
               <ProgressTableLevelBubble
                 levelStatus={subStatus}
-                isDisabled={!!level.bonus && !this.props.stageExtrasEnabled}
                 bubbleSize={progressStyles.BubbleSize.letter}
```
Comment (cforkish) [inline apps/src/templates/sectionProgress/progressTables/ProgressTableDetailCell.jsx:67]: now that `ProgressTableLevelBubble` doesn't need to support a disabled state, we no longer need the `stageExtrasEnabled` prop.

File: apps/src/code-studio/progressRedux.js [hunk 2]
```diff
@@ -556,12 +556,9 @@ const peerReviewLevels = state =>
   state.peerReviewLessonInfo.levels.map((level, index) => ({
     // These aren't true levels (i.e. we won't have an entry in levelResults),
     // so always use a specific id that won't collide with real levels
+    ...level,
```
Comment (jamescodeorg) [inline apps/src/code-studio/progressRedux.js:560]: Just checking -- is level.icon only set when level.locked is true?

File: apps/src/templates/progress/progressStyles.js [hunk 3]
```diff
@@ -232,30 +232,9 @@ function shapeSizeStyle(shape, size) {
  */
 
 /**
- * Get border and background styling based on level and student progress.
- *
- * Note: `levelProgressStyle` below is used by our legacy bubble components,
- * but once those get wrapped into our new components these two functions can
- * be merged into one.
- */
-export function getProgressStyle({
```
Comment (jamescodeorg) [inline apps/src/templates/progress/progressStyles.js:241]: Nice!

File: apps/src/templates/progress/progressStyles.js [hunk 4]
```diff
@@ -232,30 +232,9 @@ function shapeSizeStyle(shape, size) {
...
-  };
-}
-
-/**
- * Given a level object, figure out styling related to its color, border color,
- * and background color
+ * Get border and background styling based on level kind and student progress.
  */
-export const levelProgressStyle = (levelStatus, levelKind, disabled) => {
+export function levelProgressStyle(levelStatus, levelKind) {
```
Comment (jamescodeorg) [inline apps/src/templates/progress/progressStyles.js:237]: I think the intent here is that levelStatus and/or levelKind can be be undefined here, right?  If so, could you add that information to the header comment and add tests for those cases?

File: dashboard/app/helpers/users_helper.rb [hunk 5]
```diff
@@ -199,99 +199,95 @@ def script_progress_for_users(users, script)
...
-        locked: locked || nil
-      }
+  private def get_level_progress(
+    user_level,
+    script_level,
+    paired_user_levels,
+    include_timestamp
+  )
+    # if we don't have a user level, we don't need to return anything,
+    # unless the lesson is lockable, in which case that means it's locked
```
Comment (jamescodeorg) [inline dashboard/app/helpers/users_helper.rb]: Could you rephrase this comment?   It's a little bit difficult to parse.  (It might help to include _why_ we want this behavior.

File: dashboard/app/helpers/users_helper.rb [hunk 6]
```diff
@@ -199,99 +199,95 @@ def script_progress_for_users(users, script)
...
+    if user_level.nil?
+      if script_level.lesson.lockable?
+        return {locked: true}
+      else
+        return nil
+      end
     end
 
-    # a user_level is submitted if the state is submitted UNLESS it is a peer reviewable level that has been reviewed
-    submitted = !!user_level&.submitted &&
```
Comment (jamescodeorg) [inline dashboard/app/helpers/users_helper.rb:267]: We don't need to send submitted anymore?

File: dashboard/app/helpers/users_helper.rb [hunk 7]
```diff
@@ -199,99 +199,95 @@ def script_progress_for_users(users, script)
...
-      locked: locked || nil,
-      result: user_level&.best_result || 0,
-      submitted: submitted || nil,
-      readonly_answers: !!user_level&.readonly_answers || nil,
-      paired: (paired_user_levels.include? user_level&.id) || nil,
-      last_progress_at: include_timestamp ? user_level&.updated_at&.to_i : nil,
-      time_spent: user_level&.time_spent&.to_i
+      status: activity_css_class(user_level),
+      locked: user_level.show_as_locked?(script_level.lesson) || nil,
+      result: user_level.best_result || nil,
```
Comment (jamescodeorg) [inline dashboard/app/helpers/users_helper.rb:268]: Just curious, do we still use both `status` and `result` on the front-end?  (If we're close to eliminating one or the other, let's tackle in future PR.)

File: dashboard/app/models/user_level.rb [hunk 8]
```diff
@@ -141,11 +141,10 @@ def locked=(is_locked)
     self.unlocked_at = is_locked ? nil : Time.now
   end
 
-  # this is the "locked" value we return to the client. if the lesson isn't
-  # lockable or we are an authorized teacher, we always return `false`.
+  # this is the "locked" value we return to the client.
+  # if the lesson isn't lockable, we always return `false`.
   def show_as_locked?(stage)
     return false unless stage.lockable?
-    return false if user.authorized_teacher?
```
Comment (jamescodeorg) [inline dashboard/app/models/user_level.rb:148]: Where did this logic go?

**Review by jamescodeorg — COMMENTED** [review]
A few minor questions, likely no changes needed.

This is another huge step forward in untangling the progress mess, nice work!

File: apps/src/code-studio/progressRedux.js (hunk 2 above)
Comment (cforkish) [reply apps/src/code-studio/progressRedux.js:560]: yeah.

File: apps/src/templates/progress/progressStyles.js (hunk 4 above)
Comment (cforkish) [reply apps/src/templates/progress/progressStyles.js:237]: turns out we don't have any tests for this module yet. would you be okay with me deferring that work until the PR that actually consolidates our use of this stuff through `BubbleFactory` (or whatever that evolves into)?

File: dashboard/app/helpers/users_helper.rb (hunk 7 above)
Comment (cforkish) [reply dashboard/app/helpers/users_helper.rb:268]: we still use both. it will be great to someday get rid of one, but that's definitely out of scope for now.

File: dashboard/app/helpers/users_helper.rb (hunk 6 above)
Comment (cforkish) [reply dashboard/app/helpers/users_helper.rb:267]: nope! it used to serve as a sort of synonym for locked, but that has been cleaned up so it's no longer used anywhere.

File: dashboard/app/models/user_level.rb (hunk 8 above)
Comment (cforkish) [reply dashboard/app/models/user_level.rb:148]: the short answer is that we don't need it. 

one thing this work has clarified is that `locked` and `disabled` aren't necessarily synonymous. we want to show the lock icon on locked levels even for authorized teachers, but we _don't want_ to disable those bubbles for authorized teachers. the disabling part is handled on the frontend, so removing this line is just to allow us to show the lock icon to authorized teachers

**Review by jamescodeorg — APPROVED** [review]
